### ISSUE TITLE
StrMap fromFoldable stack explosion

### ISSUE THREAD

**Issue by matthewleon** [purescript/purescript-maps#108]
try
```
M.fromFoldable $ (flip Tuple) unit <<< show <$> L.range 0 9999
```

This works.

Now add a 9. Boom.

`Data.Map` maps work fine with this.

**Comment by natefaubion** [comment]
It uses `for_` which is not stack-safe. An alternative would be to fold into an `Array` first, and use `foreachE`.

**Comment by matthewleon** [comment]
> It uses for_ which is not stack-safe. An alternative would be to fold into an Array first, and use foreachE.

A rock and a hard place. Is there no balm in Gilead?

**Comment by matthewleon** [comment]
More seriously, I feel like there has to be a better way to fix this than generating a big pile of garbage to collect. But maybe I am very wrong.

**Comment by natefaubion** [comment]
You can also use unsafe effects with `foldl/foldr`.

**Comment by matthewleon** [comment]
> You can also use unsafe effects with foldl/foldr.

That's like using the cheat code and saying you beat the game, though, isn't it?

**Comment by hdgarrood** [comment]
Should this be considered a bug in the `Traversable List` instance?

**Comment by matthewleon** [comment]
> Should this be considered a bug in the Traversable List instance?

I would say so. I'm having a look at what can be done there.

**Comment by natefaubion** [comment]
> Should this be considered a bug in the Traversable List instance?

`for_` is only Foldable in any case. The unsafety is in the sequencing of Applicative effects.

**Comment by natefaubion** [comment]
> That's like using the cheat code and saying you beat the game, though, isn't it? :)

You may only use `foldl/foldr`, but then you generate garbage by copying `StrMap`s. If you want to use mutation, you must sequence effects, which is not necessarily stack safe. To guarantee that, you must use `MonadRec` (ala `purescript-safely`), which has it's own overhead. Pick two: purity, stack-safety, effeciency.

**Comment by matthewleon** [comment]
> Pick two: purity, stack-safety, effeciency.

![i-want-it-all-and-i-want-it-now](https://cloud.githubusercontent.com/assets/278859/26791191/2c2539e8-4a0e-11e7-93ec-3dafc42142c0.png)

**Comment by natefaubion** [comment]
You can also wait until proper tail-calls are implemented pervasively, and the problem will probably take care of itself.

**Comment by matthewleon** [comment]
> You can also wait until proper tail-calls are implemented pervasively

Is there an issue for this one? Did a search and couldn't find it.

**Comment by natefaubion** [comment]
I was referring to tail-call support in JS engines :)

**Comment by matthewleon** [comment]
> I was referring to tail-call support in JS engines :)

Ah yes, indeed. How will that help with `foldr` though?

**Comment by michaelficarra** [comment]
Unfortunately, Safari and V8-behind-a-flag will be the only implementations for a while. Edge has inherent call stack design limitations which prevent it and Firefox has inherent security design limitations which prevent it.

**Comment by hdgarrood** [comment]
Was this fixed by #110?

**Comment by natefaubion** [comment]
Yes.
